### ISSUE TITLE
GasPriceLowerThanBaseFee errors on eth_call

### ISSUE THREAD

**Issue by KaiRo-at** [openethereum/openethereum#446]
- **OpenEthereum version (>=3.1.0)**: 3.3.0-rc2
- **Operating system**: Linux
- **Installation**: docker
- **Fully synchronized**: yes
- **Network**: ropsten
- **Restarted**: yes

When doing `eth_call` on Ropsten, now that London is enabled, I always get a GasPriceLowerThanBaseFee error, see e.g. this direct JSON-RPC call via `curl`:

```
> curl -H "Content-Type: application/json" --data '{"jsonrpc":"2.0","method":"eth_call","params":[{"to": "0x588Ca95d76c6a7ab121746bd209F2994B7Ba5F10", "data": "0x63a6344e"}, "latest"],"id":67}' http://localhost:8545
{"jsonrpc":"2.0","error":{"code":-32015,"message":"Transaction execution error.","data":"GasPriceLowerThanBaseFee { gas_price: 0, base_fee: 159911300881 }"},"id":67}
```

When I do this call on Infura instead, it works as expected:

```
> curl -H "Content-Type: application/json" --data '{"jsonrpc":"2.0","method":"eth_call","params":[{"to": "0x588Ca95d76c6a7ab121746bd209F2994B7Ba5F10", "data": "0x63a6344e"}, "latest"],"id":67}' https://ropsten.infura.io/v3/...
{"jsonrpc":"2.0","id":67,"result":"0x0000000000000000000000000000000000000000000000000000000000000000"}
```

IIRC, this was discussed in the recent ACD call, and there seems to be a link for a discussion on the topic: https://github.com/ethereum/pm/issues/330#issuecomment-854773045

**Comment by dwhjames** [comment]
Related to this, [`trace_rawTransaction`](https://openethereum.github.io/JSONRPC-trace-module#trace_rawtransaction) exhibits the same error:
```
GasPriceLowerThanBaseFee { gas_price: 1, base_fee: 491224310953 }
```

**Comment by salisbury-espinosa** [comment]
need to add "gasPrice"
request:

`curl -H "Content-Type: application/json" --data '{"jsonrpc":"2.0","method":"eth_call","params":[{"gasPrice":"0x10000", "to": "0x588Ca95d76c6a7ab121746bd209F2994B7Ba5F10", "data": "0x63a6344e"}, "latest"],"id":67}' http://localhost:8080`

response:

`{"jsonrpc":"2.0","result":"0x0000000000000000000000000000000000000000000000000000000000000000","id":67}`

**Comment by KaiRo-at** [comment]
> try to add "gasPrice"

Of course that works as a workaround. But that's not what libraries do or should need to do. The `curl` example I put in the initial comment is just for illustration of the problem.

**Comment by yuga-cb** [comment]
FWIW, I am seeing that this can fail even if `gas_price` is set to something twice as large as the most recent base fee.

`RPC call failed: {"code"=>-32015, "message"=>"Transaction execution error.", "data"=>"GasPriceLowerThanBaseFee { gas_price: 2935274, base_fee: 240650928267 }"}`

**Comment by sunce86** [comment]
> FWIW, I am seeing that this can fail even if `gas_price` is set to something twice as large as the most recent base fee.
> 
> `RPC call failed: {"code"=>-32015, "message"=>"Transaction execution error.", "data"=>"GasPriceLowerThanBaseFee { gas_price: 2935274, base_fee: 240650928267 }"}`

Sorry, but I might be missing something, but how 2935274 is twice as large as 240650928267 ?

**Comment by sunce86** [comment]
> Related to this, [`trace_rawTransaction`](https://openethereum.github.io/JSONRPC-trace-module#trace_rawtransaction) exhibits the same error:
> 
> ```
> GasPriceLowerThanBaseFee { gas_price: 1, base_fee: 491224310953 }
> ```

Hi, 
So, for trace_rawTransaction you can't omit to specify gas_price, and if you specify gas_price then it must be higher then current base_fee. 
Please refer to explanation here: https://github.com/ethereum/go-ethereum/pull/23027
In short words, for all RPC functions that work with execution of transactions, there are two options:
1. You don't specify gas prices => base fee is forced to zero therefore ignored
2. You specify gas prices => gas prices are validated against current base fee and it must be valid gas prices >= base fee

**Comment by yuga-cb** [comment]
> > FWIW, I am seeing that this can fail even if `gas_price` is set to something twice as large as the most recent base fee.
> > `RPC call failed: {"code"=>-32015, "message"=>"Transaction execution error.", "data"=>"GasPriceLowerThanBaseFee { gas_price: 2935274, base_fee: 240650928267 }"}`
> 
> Sorry, I might be missing something, but how 2935274 is twice as large as 240650928267 ?

For this call, we are setting gas_price equal to 2 * the base fee of the _most recent_ block (1467637 * 2 = 2935274). The node seems to think that the base_fee is 240650928267. This could be because we are specifying a block_number in eth_call, and maybe the node is looking at the base_fee of that block_number rather than the most recent block. Regardless, this seems like a bug.

**Comment by sunce86** [comment]
> > > FWIW, I am seeing that this can fail even if `gas_price` is set to something twice as large as the most recent base fee.
> > > `RPC call failed: {"code"=>-32015, "message"=>"Transaction execution error.", "data"=>"GasPriceLowerThanBaseFee { gas_price: 2935274, base_fee: 240650928267 }"}`
> > 
> > 
> > Sorry, I might be missing something, but how 2935274 is twice as large as 240650928267 ?
> 
> For this call, we are setting gas_price equal to 2 * the base fee of the _most recent_ block (1467637 * 2 = 2935274). The node seems to think that the base_fee is 240650928267. This could be because we are specifying a block_number in eth_call, and maybe the node is looking at the base_fee of that block_number rather than the most recent block. Regardless, this seems like a bug.

Yes, the node is always looking at the base fee of the specified block. That is expected.
Anyway, let's revise all this cases once we have the fix.

**Comment by sunce86** [comment]
> FWIW, I am seeing that this can fail even if `gas_price` is set to something twice as large as the most recent base fee.
> 
> `RPC call failed: {"code"=>-32015, "message"=>"Transaction execution error.", "data"=>"GasPriceLowerThanBaseFee { gas_price: 2935274, base_fee: 240650928267 }"}`

Did you find time to check this issue with the newest release? Can we close this issue?

**Comment by KaiRo-at** [comment]
FWIW, my original issue is fixed with OpenEthereum 3.3.0-rc.3 which I have been using on both Ropsten and Görli since it was released.

**Comment by sunce86** [comment]
PR merged. Closed.
